### ISSUE TITLE
Streaming Subscriptions missing data when batch is smaller than amount of relevant rows

### ISSUE THREAD

**Issue by Bessonov** [hasura/graphql-engine#9528]
### Version Information

Server Version: v2.21.0
CLI Version (for CLI related issue):

### Environment

OSS

### What is the current behaviour?

An entity have `updatedat timestamptz(3)` column, which is used to show when it was updated. The column is managed by before insert or update trigger. The test table contains two relevant rows.

The subscription:
```graphql
subscription organizationsChanges {
  organizations_stream(cursor: {initial_value: {updatedat: "2000-01-01"}, ordering: ASC}, batch_size: 1) {
    id
    name
  }
}
```
initially returns only one row, even though it should return two rows in the test case. Furthermore, if the rows are updated in the same transaction and therefore have the same timestamp, only one row is returned.

### What is the expected behaviour?

In my opinion, in this case the streaming subscription should return two batches of one row each.

### How to reproduce the issue?

1. Create a table with a timestamp or number and insert some rows.
2. Subscribe to the events with a batch smaller than the number of rows -> not all rows above the initial value will be returned.
3. Increase number or timestamp to all rows at once -> not all relevant rows are returned

### Screenshots or Screencast

-

### Please provide any traces or logs that could help here.

-

### Any possible solutions/workarounds you're aware of?

Increasing the batch size helps for some use cases, but not for software like task management and chat, which support mass data edits.

Another workaround is to stick with the non-streaming subscription.

### Keywords

stream streaming subscription batch cursor

**Comment by Bessonov** [comment]
I noticed that if the `updatedat` values are different, then it works as expected in both cases, the initial load and on changes. To check for changes, I have dropped the trigger and set the date manually:

```sql
update organizations as t set
    updatedat = c.updatedat
from (values
    ('13f5625a-df89-45fb-b954-d8591d68a01b', now()),
    ('9910164e-37a1-455f-880d-4d3d1cef994c', now() - interval '1 second')  
) as c(id, updatedat) 
where c.id::uuid = t.id;
```

It's probably not a bug, but a feature request. I found a trace in the [documentation](https://hasura.io/docs/latest/subscriptions/postgres/streaming/index/#how-it-works):

> Ideally, the cursor chosen should represent unique

I think "ideally" isn't the ideal term.

**Comment by michaelAtCoalesce** [comment]
so is the issue here that where inserting that the update values are the same?

**Comment by Bessonov** [comment]
@michaelAtCoalesce AFAIR, the number of rows returned within each batch is correct. However, there aren't enough batches returned, so the total number of rows returned is not correct.

@adas98012 Thanks for fixing the labels. However, I don't believe that the solution should be limited to just improving the documentation.

**Comment by michaelAtCoalesce** [comment]
im experimenting a little and instead of using an update timestamp, which, as mentioned above, does appear to be identical across multiple rows if you insert them at the same time..... i'm using a column with a sequence and using nextval('Sequence') whenever im inserting or updating. ultimately, i think the idea for the updated timestamp is to keep track of a write ordering, and a sequence where all inserts and updates use nextval() serves the same purposes, but it does assign unique IDs when inserting multiple rows

**Comment by Bessonov** [comment]
@michaelAtCoalesce yes, it should work. I am using the timestamp because I need it anyway.

**Comment by Zerebokep** [comment]
Batching has its flaws when you use timestamptz as cursor, you should switch to an integer (e.g. your auto incremental pk).

**Comment by Bessonov** [comment]
@Zerebokep Note that I am not interested in getting new entries, but updated entries.

**Comment by Zerebokep** [comment]
I see, then you need to add a monotonic id which also increases on UPDATE's.
